gordon_janitor_gcp/gce_clients.py

```diff
@@ -0,0 +1,297 @@
+
+To use:
+
+.. code-block:: python
+
+    import asyncio
+
+    import aiohttp
+
+    from gordon_janitor_gcp import auth
+
+    async def main():
+        session = aiohttp.ClientSession()
+        loop = asyncio.get_event_loop()
+        auth_client = auth.GoogleAuthClient(k
+            eyfile='/path/to/keyfile', session=session)
+        client = GCPClient(auth_client, session)
+        instances = await client.list_instances('project-id')
+        print(instances)
+
+    loop.run_until_complete(main())
+    # example output
+    # [{'hostname': 'instance-1', 'internal_ip': '10.10.10.10',
+    #   'external_ip': '192.168.1.10'}]
+"""
+
+import asyncio
+import logging
+
+from gordon_janitor_gcp import exceptions
+from gordon_janitor_gcp import http_client
+
+
+class PagingGoogleClientBase(http_client.AIOGoogleHTTPClient):
+    """HTTP client base class that extracts data from paginated Google APIs.
+
+    Args:
+        auth_client (gordon_janitor_gcp.auth.GoogleAuthClient): client
+            to manage authentication for HTTP API requests.
+        session (aiohttp.ClientSession): (optional) ``aiohttp`` HTTP
+            session to use for sending requests. Defaults to the
+            session object attached to ``auth_client`` if not provided.
+    """
+
+    def __init__(self, auth_client=None, session=None):
+        super().__init__(auth_client=auth_client, session=session)
+
+    def parse_response_into_items(self, response, items):
+        """Parse an API response and append the results to ``items``
+
+        Must be implemented by inheriting class.
+
+        Args:
+            response (dict): API json response.
+            items (list): list for collecting parsed results.
+        """
+        raise NotImplementedError(
+            f'{self.__class__.__name__} implement "parse_response_into_items" '
+            'method!')
+
+    async def list_all_items(self, url, params, retries, retry_wait):
+        """Aggregate data from all pages of an API query.
+
+        Args:
+            url (str): Google API endpoint URL.
+            params (dict): URL query parameters.
+            retries (int): Number of times to retry an API call.
+            retry_wait (int): Number of seconds to wait between each retry.
+        Returns:
+            list: parsed query response results.
+        """
+        items = []
+        next_page_token = None
+
+        while True:
+            if next_page_token:
+                params['pageToken'] = next_page_token
+            try:
+                response = await self.get_json(url, params=params)
+            except exceptions.GCPHTTPError as e:
+                base_msg = f'Fetching items from "{url}" failed: {e}, '
+                if retries > 0:
+                    retries -= 1
+                    msg = base_msg + f'RETRYING. {retries} retries left.'
+                    logging.error(msg)
+                    await asyncio.sleep(retry_wait)
+                    continue
```

Comment:
When continuing, the execution jumps back to the beginning of the loop, so `self.parse_response_into_items` won't get called. If a response object is never created because all the attempts fail, the function will return in the `else` clause.